### DISCUSSION TITLE
Saving Session Automation

### DISCUSSION THREAD

**Discussion by funks81** [BottlecapDave/HomeAssistant-OctopusEnergy#534] [General]
Has anyone created an automation that they're happy to share for automating battery charging prior to a saving session and then discharge during the saving session.  My coding is particularly poor - so was wondering if someone could save me the brainache of trying to automate this!!!

**Comment by Gary-Manuel** [comment]
Yes this works perfectly https://www.facebook.com/groups/615579009972782/permalink/859319558932058 
and 
https://www.facebook.com/groups/615579009972782/permalink/867158271481520

**Reply by funks81** [reply]
Looks good - may require a fair bit of work to get it to work with my Solis inverter

**Comment by GuitaristForEver** [comment]
## Response to @rickbrouwer (Static Fallback Limitation)

You're absolutely right - when **all triggers fail**, we still use static replicas. This is **risk reduction**, not elimination.

### What Changes

**Before**: Single trigger fails → 100% chance of static fallback
**After**: Static fallback only when ALL triggers fail simultaneously

If primary has 99% uptime and secondary (CPU) has 99.9% uptime → combined failure drops from 1% to 0.001%. That's **1000x risk reduction**.

### N-Tier Failover

The formula approach supports multiple tiers:

```yaml
triggers:
  - type: metrics-api    # Custom metrics
  - type: prometheus     # Cluster metrics
  - type: cpu           # Kernel metrics (almost never fails)

formula: "primary ?? secondary ?? tertiary ?? 5"
```

Static replicas become the **last resort** after 3+ independent systems fail, not the first resort.

### What This Solves

✅ **High availability for single-trigger failures** (the common case) - maintains dynamic scaling
✅ **Cost optimization** - no over-provisioning during partial outages
✅ **Progressive degradation** - graceful fallback chain instead of immediate static

### Why Not Eliminate Static Fallback?

Alternatives are worse:
- Scale to max? Resource exhaustion risk
- Pause at current? Already exists as `behavior: paused`, doesn't help if load spikes
- Historical data? Complex, potentially stale

### Bottom Line

This transforms the problem from **"every primary failure = static"** to **"only N simultaneous failures = static"**. It's pragmatic and significantly better than the status quo.

For ultra-critical systems: use highly reliable secondaries (CPU/memory), add 3rd tier, set conservative `fallback.replicas` for "everything broken" scenarios.
